File: 2/matrix/MatrixLowTriangle.cpp
#include <stdexcept>
#include "MatrixLowTriangle.h"

MatrixLowTriangle::MatrixLowTriangle(int _size) : Matrix(_size) {
    data = new double[getArraySize()];
}

MatrixLowTriangle::MatrixLowTriangle(const MatrixLowTriangle &other) : Matrix(other.size) {
    data = new double[getArraySize()];
}

MatrixLowTriangle &MatrixLowTriangle::operator=(const MatrixLowTriangle &other) {
    if (size != other.size) {
        delete data;
        data = new double[other.getArraySize()];
    }
    size = other.size;
    for (int i = 0; i < getArraySize(); i++) {
        data[i] = other.data[i];
    }
    return *this;
}

MatrixLowTriangle::~MatrixLowTriangle() {
    delete data;
    data = nullptr;
}

double MatrixLowTriangle::get(int i, int j) {
    if (i < 0 || j < 0 || i >= size || j >= size) {
        throw std::out_of_range("Matrix out of range");
    } else if (j <= i) {
        return data[getCellIndex(i, j)];
    } else {
        return 0;
    }
}

void MatrixLowTriangle::put(int i, int j, double value) {
    if (i < 0 || j < 0 || i >= size || j >= size) {
        throw std::out_of_range("Matrix out of range");
    } else if (j <= i) {
        data[getCellIndex(i, j)] = value;
    } else {
        throw std::invalid_argument("Can't set value above main diagonal");
    }
}


int MatrixLowTriangle::getArraySize() const{
    return (size * (size + 1)) / 2;
}

int MatrixLowTriangle::getCellIndex(int i, int j) {
    int passed = (i * (i + 1)) / 2;
    return passed + j;
}

File: 3/matrix/MatrixLowTriangle.py
from matrix import Matrix


class MatrixLowTriangle(Matrix):
    def __init__(self, size):
        super().__init__(size)
        self.data = [0] * self._get_array_size()

    def get(self, i, j):
        self._index_check(i, j)
        if j <= i:
            return self.data[self._get_cell_index(i, j)]
        else:
            return 0

    def put(self, i, j, value):
        self._index_check(i, j)
        if j <= j:
            self.data[self._get_cell_index(i, j)] = value
        else:
            raise IndexError("Can't set value above main diagonal")

    def _index_check(self, i, j):
        if i < 0 or i >= self.size or j < 0 or j >= self.size:
            raise IndexError("Invalid indexes")

    def _get_array_size(self):
        return (self.size * (self.size + 1)) // 2

    def _get_cell_index(self, i, j):
        passed = (i * (i + 1)) // 2
        return passed + j
